src/main/scala/nl/vroste/zio/kinesis/client/DynamicConsumer.scala

```diff
@@ -110,13 +112,104 @@ object DynamicConsumer {
+   * @param workerIdentifier Identifier used for the worker in this application group. Used in logging
+   *   and written to the lease table.
+   * @param maxShardBufferSize The maximum number of records per shard to store in a queue before blocking
+   *   the KCL record processor until records have been dequeued. Note that the stream returned from this
+   *   method will have internal chunk buffers as well.
+   * @param checkpointBatchSize Maximum number of records before checkpointing
+   * @param checkpointDuration Maximum interval before checkpointing
+   * @param recordProcessor A function for processing a `Record[T]`
+   * @tparam R ZIO environment type required by the `deserializer` and the `recordProcessor`
+   * @tparam T Type of record values
+   * @return ZIO of unit
+   */
+  def consumeWith[R, T](
+    streamName: String,
+    applicationName: String,
+    deserializer: Deserializer[R, T],
+    requestShutdown: UIO[Unit] = UIO.never,
+    initialPosition: InitialPositionInStreamExtended =
+      InitialPositionInStreamExtended.newInitialPosition(InitialPositionInStream.TRIM_HORIZON),
+    isEnhancedFanOut: Boolean = true,
+    leaseTableName: Option[String] = None,
+    workerIdentifier: String = UUID.randomUUID().toString,
+    maxShardBufferSize: Int = 1024, // Prefer powers of 2
+    checkpointBatchSize: Long = 200,
+    checkpointDuration: Duration = 5.second
+  )(
+    recordProcessor: RecordProcessor[R, T]
+  ): ZIO[R with Blocking with Logging with Clock with DynamicConsumer, Throwable, Unit] =
+    for {
+      consumer <- ZIO.service[DynamicConsumer.Service]
+      _        <- consumer
+             .shardedStream(
+               streamName,
+               applicationName,
+               deserializer,
+               requestShutdown,
+               initialPosition,
+               isEnhancedFanOut,
+               leaseTableName,
+               workerIdentifier,
+               maxShardBufferSize
+             )
+             .flatMapPar(Int.MaxValue) {
+               case (_, shardStream, checkpointer) =>
+                 ZStream.fromEffect {
+                   ZIO.environment[R].flatMap { env =>
+                     for {
+                       refSkip <- Ref.make(false)
+                       _       <- shardStream
+                              .tap(record =>
+                                processWithSkipOnError(refSkip)(
+                                  recordProcessor(record).provide(env) *> checkpointer.stage(record)
+                                )
+                              )
+                              .via(
+                                checkpointer
+                                  .checkpointBatched[Blocking with Logging](
+                                    nr = checkpointBatchSize,
+                                    interval = checkpointDuration
+                                  )
+                              )
+                              .runDrain
```

Comment:
I don't think you need the inner `runDrain` here, if you change `ZStream.fromEffect` to `ZStream.unwrap`

Reply:
:+1:

Reply:
I tried your suggestion but this leads to a compile error whereby the `shardStream` returns a `ZStream` inside a `ZIO` flatmap/for-comp